src/qt/trafficgraphwidget.cpp

```diff
@@ -1,257 +1,305 @@
-        nSamplesTailIndex += MAX_SAMPLES;
-    }
-
-    updateDisplay();
-}
-
-//MLUMIN 05_22: Instead of just clearing in and out, zero out in/out data samples, in/out display samples, if desired
-void TrafficGraphWidget::clear()
-{
-    vSamplesIn.fill(0.0f);
-    vSamplesOut.fill(0.0f);
-    vSamplesToDisplayIn.fill(0.0f);
-    vSamplesToDisplayOut.fill(0.0f);
-    fMax = 0.0f;
-
-    if (clientModel) {
-        nLastBytesIn = clientModel->getTotalBytesRecv();
-        nLastBytesOut = clientModel->getTotalBytesSent();
-
-        updateDisplay();
-    }
-}
+// Copyright (c) 2011-2015 The Bitcoin Core developers
+// Copyright (c) 2022 The Dogecoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "trafficgraphwidget.h"
+#include "clientmodel.h"
+#include "utiltime.h"
+
+#include <QPainter>
+#include <QPainterPath>
+#include <QColor>
+#include <QTimer>
+
+#include <cmath>
+
+#define NUM_DISPLAYED_SAMPLES         800
+#define BASE_SAMPLES_BUF_SIZE         24 * 60 * 60
+#define BASE_SAMPLE_PERIOD_MS         1000
+
+#define XMARGIN                 10
+#define YMARGIN                 10
+
+TrafficGraphWidget::TrafficGraphWidget(QWidget *parent) :
+    QWidget(parent),
+    fMax(0.0f),
+    nMins(0),
+
+    vTimespanSamplesIn(NUM_DISPLAYED_SAMPLES),
+    vTimespanSamplesOut(NUM_DISPLAYED_SAMPLES),
+    nTimespanLastBytesIn(0),
+    nTimespanLastBytesOut(0),
+    nTimespanLastSampleTimeMs(0),
+    nTimespanHeadIndex(0),
+    timespanTimer(0),
+    vBaseSamplesIn(BASE_SAMPLES_BUF_SIZE),
+    vBaseSamplesOut(BASE_SAMPLES_BUF_SIZE),
+    nBaseLastBytesIn(0),
+    nBaseLastBytesOut(0),
+    nBaseLastSampleTimeMs(0),
+    nBaseSamplesHeadIndex(0),
+    baseTimer(0),
+
+    clientModel(0)
+{
+    timespanTimer = new QTimer(this);
+    connect(timespanTimer, SIGNAL(timeout()), SLOT(updateTimespanRates()));
+
+    baseTimer = new QTimer(this);
+    connect(baseTimer, SIGNAL(timeout()), SLOT(updateBaseRates()));
+}
+
+void TrafficGraphWidget::setClientModel(ClientModel *model)
+{
+    clientModel = model;
+    if(model) {
+        nTimespanLastBytesIn = model->getTotalBytesRecv();
+        nTimespanLastBytesOut = model->getTotalBytesSent();
+        nBaseLastBytesIn = nTimespanLastBytesIn;
+        nBaseLastBytesOut = nTimespanLastBytesOut;
+
+        baseTimer->start(BASE_SAMPLE_PERIOD_MS);
+    }
+}
+
+int TrafficGraphWidget::getGraphRangeMins() const
+{
+    return nMins;
+}
+
+float TrafficGraphWidget::resampleSamples(const QVector<float> &source, QVector<float> &dest) const
+{
+    int numSourceSamples = nMins * 60;
+    float stride = (float)numSourceSamples / dest.size();
+    float currentStride = stride;
+    float maxSample = 0.0f;
+    float lastAverageRate = 0.0f;
+
+    int sourceTailIndex = nBaseSamplesHeadIndex - numSourceSamples;
+    if (sourceTailIndex < 0) {
+        sourceTailIndex += BASE_SAMPLES_BUF_SIZE;
+    }

```

Comment:
I'm trying to figure out what this represents without re-reading the previous discussion; is this an index into a ring buffer, so that this addition if negative achieves wraparound behavior?

If so, a comment to that effect could make this clearer.

Reply:
I have a different idea; let me resolve this and comment elsewhere.